src/awkward/operations/ak_from_iter.py

```diff
@@ -99,6 +99,13 @@ def _impl(iterable, highlevel, behavior, allow_record, initial, resize):
     formstr, length, buffers = builder.to_buffers()
     form = ak.forms.from_json(formstr)
 
-    return ak.operations.from_buffers(
-        form, length, buffers, highlevel=highlevel, behavior=behavior
+    return ak.operations.ak_from_buffers._impl(
+        form,
+        length,
+        buffers,
+        buffer_key="{form_key}-{attribute}",
+        backend="cpu",
+        highlevel=highlevel,
+        behavior=behavior,
+        simplify=True,
```

Comment:
Needed here because of ArrayBuilder making non-canonical Forms.

Halfway through implementing this, I was wondering if maybe were being too aggressive with defining option-of-union as non-canonical, when we use it everywhere. However, this PR demonstrates that it's possible to remove all cases of it, it is redundant with union-of-options, and Arrow has set a convention that we might as well use.